File: iOS/Kalorie/Core/Utils/NutritionLabelPrefilling.swift
//
//  NutritionLabelPrefilling.swift
//  Kalorie
//
//  Created by Josef Antoni on 13.09.2026.
//

import Foundation
import UIKit

protocol NutritionLabelPrefilling: AnyObject {
    var formInput: FoodItemFormInput { get set }
    var recognizedFields: Set<FoodItemFormField> { get set }
    var isRecognizingNutritionLabel: Bool { get set }
    var isNutritionLabelCameraVisible: Bool { get set }
    var nutritionLabelCameraHint: String? { get set }
    var alertItem: AlertItem? { get set }
}

extension NutritionLabelPrefilling {

    @MainActor
    @discardableResult
    func recognizeNutritionLabel(
        from image: UIImage,
        liveBarcode: String?,
        using recognizeNutritionLabel: any RecognizeNutritionLabelUseCaseProtocol
    ) async -> Bool {
        guard !isRecognizingNutritionLabel else { return false }
        isRecognizingNutritionLabel = true
        defer { isRecognizingNutritionLabel = false }
        do {
            var reading = try await recognizeNutritionLabel(image: image)
            reading.scannedCode = reading.scannedCode ?? liveBarcode
            guard !reading.recognizedFields.isEmpty else {
                nutritionLabelCameraHint = L10n.AddFood.nutritionLabelNothingRecognized
                return false
            }
            recognizedFields.formUnion(formInput.applying(reading, alreadyRecognizedFields: recognizedFields))
            nutritionLabelCameraHint = nil
            isNutritionLabelCameraVisible = false
            return true
        } catch NutritionLabelRecognitionError.nothingRecognized {
            nutritionLabelCameraHint = L10n.AddFood.nutritionLabelNothingRecognized
            return false
        } catch {
            Log.warning(error, category: Constants.LogCategory.nutritionLabelRecognition)
            nutritionLabelCameraHint = L10n.AddFood.nutritionLabelNothingRecognized
            return false
        }
    }

    @MainActor
    func openNutritionLabelCamera(using cameraAuthorizationProvider: any CameraAuthorizationProviderProtocol, onDenied: () -> Void = {}) async {
        switch cameraAuthorizationProvider.status {
        case .authorized:
            isNutritionLabelCameraVisible = true
        case .notDetermined:
            if await cameraAuthorizationProvider.requestAccess() {
                isNutritionLabelCameraVisible = true
            } else {
                onDenied()
            }
        case .denied, .unsupported:
            onDenied()
        }
    }

    func onFormFieldEdited(_ field: FoodItemFormField) {
        recognizedFields.remove(field)
    }
}

File: Android/app/src/main/java/antoni/kalorie/core/utils/NutritionLabelPrefilling.kt
package antoni.kalorie.core.utils

import antoni.kalorie.R
import antoni.kalorie.components.FoodItemFormField
import antoni.kalorie.core.nutritionlabelrecognition.NutritionLabelImage
import antoni.kalorie.core.nutritionlabelrecognition.NutritionLabelRecognitionError
import antoni.kalorie.core.nutritionlabelrecognition.RecognizeNutritionLabelUseCaseProtocol
import antoni.kalorie.features.addfoodsheet.FoodItemFormInput
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow

interface NutritionLabelPrefilling {
    val formInput: MutableStateFlow<FoodItemFormInput>
    val recognizedFields: MutableStateFlow<Set<FoodItemFormField>>
    val isRecognizingNutritionLabel: MutableStateFlow<Boolean>
    val isNutritionLabelCameraVisible: MutableStateFlow<Boolean>
    val nutritionLabelCameraHintRes: MutableStateFlow<Int?>

    // MARK: - Functions

    suspend fun recognizeNutritionLabel(
        image: NutritionLabelImage,
        liveBarcode: String?,
        useCase: RecognizeNutritionLabelUseCaseProtocol,
    ): Boolean {
        if (isRecognizingNutritionLabel.value) return false
        isRecognizingNutritionLabel.value = true
        try {
            val recognized = useCase(image)
            val reading = recognized.copy(scannedCode = recognized.scannedCode ?: liveBarcode)
            if (reading.recognizedFields.isEmpty()) {
                nutritionLabelCameraHintRes.value = R.string.addFood_nutritionLabel_nothingRecognized
                return false
            }
            val (applied, appliedFields) = formInput.value.applying(reading, recognizedFields.value)
            formInput.value = applied
            recognizedFields.value = recognizedFields.value + appliedFields
            nutritionLabelCameraHintRes.value = null
            isNutritionLabelCameraVisible.value = false
            return true
        } catch (error: CancellationException) {
            throw error
        } catch (_: NutritionLabelRecognitionError.NothingRecognized) {
            nutritionLabelCameraHintRes.value = R.string.addFood_nutritionLabel_nothingRecognized
            return false
        } catch (error: Exception) {
            Log.warning(error, Constants.LogCategory.NUTRITION_LABEL_RECOGNITION)
            nutritionLabelCameraHintRes.value = R.string.addFood_nutritionLabel_nothingRecognized
            return false
        } finally {
            isRecognizingNutritionLabel.value = false
        }
    }

    fun onFormFieldEdited(field: FoodItemFormField) {
        recognizedFields.value = recognizedFields.value - field
    }
}
